README.md
# Juttle Engine

[![Build Status](https://travis-ci.org/juttle/juttle-engine.svg?branch=master)](https://travis-ci.org/juttle/juttle-engine)

Juttle Engine is an integrated environment for developing and executing
[juttle](https://github.com/juttle/juttle) programs and visualizations. It lets
you run Juttle programs stored on the file system and present the results in a
browser for experimentation, development, deployment, and debugging of juttle
programs.

![](docs/screenshots/stock_prices_example.png)

To make it easy to interconnect analytics with visualization,
Juttle promotes a tight coupling between the language and the client-side
visualization library.  Rather than
having to tie together by hand a data processing layer with a separate
visualization layer, Juttle enables you to integrate these two layers
so you don't have to worry about the details
of wiring Juttle dataflow computation to your browser-based views.
Juttle Engine is the development and execution environment in which you can
explore your data using Juttle and deploy rich visualizations into production.

Under the covers, Juttle Engine integrates
[juttle-service](https://github.com/juttle/juttle-service), a node.js API server
that enables execution of Juttle programs using a REST API with the ability to
serve web application bundles. It comes with the
[juttle-viewer](https://github.com/juttle/juttle-viewer) application that
provides a simple in browser experience for loading and executing juttle
programs in the juttle-engine and rendering the results using the
[juttle-viz](https://github.com/juttle/juttle-viz) visualization library. In the
future, Juttle Engine may be extended to support other application bundles.

## Getting Started

### Installation

Make sure you have [node](http://nodejs.org) (with [npm](http://npmjs.org)) installed.

Use npm to install Juttle Engine
```
$ npm install juttle-engine
```

This will install the Juttle Engine daemon and client binaries, the [juttle](http://github.com/juttle/juttle) interpreter, and the current set of supported juttle adapters.

We believe it to work with node.js 4.2.3 and 5.0, npm 2.14.17 and 3.5, and likely other combinations we have not tested against.

### Running your first juttle program

Start the daemon by running juttle-engine:
```
$ juttle-engine -d -o juttle-engine.log
```
This will daemonize juttle-engine and output logs to juttle-engine.log. You can now run juttle programs against the juttle-engine daemon via the juttle-engine-client:
```
$ juttle-engine-client browser --path examples/core-juttle/hello_world.juttle
```
This will open a browser window and display the output of the program. You can make edits to your juttle, save the file, and reload the browser window to get the updated output.

### Examples

Several demo-style examples of Juttle programs with sample data sets are available in this repo under [examples](./examples/README.md). The same examples are deployed on a demo system [demo.juttle.io](http://demo.juttle.io/index.html).

We also offer a [tutorial](http://juttle.github.io/juttle/concepts/juttle_tutorial) for developers new to Juttle.

## Ecosystem

Here's how the juttle-engine module fits into the overall [Juttle Ecosystem](https://github.com/juttle/juttle/blob/master/docs/juttle_ecosystem.md):

[![Juttle Ecosystem](https://github.com/juttle/juttle/raw/master/docs/images/JuttleEcosystemDiagram.png)](https://github.com/juttle/juttle/blob/master/docs/juttle_ecosystem.md)

## Options and Configuration

Here are the full command line options supported by the daemon and client programs:

### juttle-engine

```
usage: [--port <port>] [--root <path>]
       [--config <juttle-config-path>] [--daemonize]
       [--output <logfile>] [--log-config <log4js-config-path>]
       [--log-level <level] [--help]
       -p, --port <port>:                     Run juttle-engine on specified port
       -r, --root <path>:                     Use <path> as the root directory for juttle programs
       -c, --config <juttle-config-path>:     Read juttle config from <juttle-config-path>
       -d, --daemonize:                       Daemonize juttle-engine and log to configured log file
       -o, --output <logfile>:                Log to specififed file when daemonized
       -L, --log-config <log4js-config-path>: Configure logging from <log4js-config-path>. Overrides any value of -o
       -l, --log-level <level>:               Use a default log level of <level>. Overridden by any log level specified in -L
       -h, --help:                            Print this help and exit');
```

``juttle-engine`` uses log4js for logging and by default logs to ``log/juttle-engine.log``.

### juttle-engine-client

```
usage: [--juttle-engine <hostname:port>] [--help] [COMMAND] [OPTIONS]
     [COMMAND]: one of the following, with the following options:
          subscribe (--job <job-id> | --observer <observer-id>)
          list_jobs [--job <job-id>]
          list_observers
          run --path <path-to-juttle-file> [--wait] [--observer <observer-id>]
          delete --job <job-id>
          get_inputs --path <path-to-juttle-file> --input name=val [--input name=val ...]
          push --path <path-to-juttle-file> [--topic <rendezvous-topic>]
          watch --path <path-to-juttle-file> [--topic <rendezvous-topic>]
          browser --path <path-to-juttle-file>
     [OPTIONS]: one of the following:
          --path <path-to-juttle-file>                 Path to file relative to configured root directory.
                                                       used by: run,get_inputs,push,watch,browser
          --wait                                       If true, wait for program to finish, otherwise run in background
                                                       used by: run
          --input name=val                             One or more input values.
                                                       used by: get_inputs
          --job <job-id>                               Job id.
                                                       used by: subscribe,list_jobs,delete
          --observer <observer-id>                     Observer id.
                                                       used by: subscribe,run
          --juttle-engine <hostname:port>              Hostname/port of server
          --help                                       Print this help and exit
          --topic <rendezvous-topic>                   Rendezvous topic
                                                       used by: push,watch
```

### Juttle config file

The Juttle compiler and runtime within juttle-engine are also configured via the juttle configuration file, typically at ``$(HOME)/.juttle/config.json``. For more information on the juttle configuration file, see the [juttle configuration documentation](https://github.com/juttle/juttle/blob/master/docs/reference/cli.md#configuration).

### Module resolution

For information on how modules are resolved, see the "Importing Modules" section in the general [juttle documentation](https://github.com/juttle/juttle/blob/master/docs/concepts/programming_constructs.md#importing-modules) on programming constructs.

## Testing

To run unit tests:

``gulp test``

To check code style and perform lint checks:

``gulp lint``

Both are run automatically by Travis.


CHANGELOG.md
# Change Log
This file documents all notable changes to juttle-engine. The release numbering uses [semantic versioning](http://semver.org).

## 0.6.0
Released 2016-03-25

### Major Changes
- Add a new [adapter](https://github.com/juttle/juttle-googleanalytics-adapter) for [Google Analytics](http://www.google.com/analytics/).

### Minor Changes
- Update to juttle version 0.7.0 [[#76](https://github.com/juttle/juttle-engine/pull/97)]
- Additional automated tests for sql adapters. [[#86](https://github.com/juttle/juttle-engine/pull/86)]
- Reduce verbosity of `npm install` step when building docker images. [[#102](https://github.com/juttle/juttle-engine/pull/102)]
- Change how modules are resolved to reflect changes in Juttle 0.7.0. Juttle programs should import modules using relative pathnames which are resolved relative to the file doing the import. [[#513](https://github.com/juttle/juttle/issues/513)]

### Bug Fixes
- Ensure that in unit tests juttle-engine is completely started before running any tests. [[#98](https://github.com/juttle/juttle-engine/pull/98)]
- Properly test that the `juttle-engine` binary can be run successfully. [[#97](https://github.com/juttle/juttle-engine/pull/97)]

## 0.5.0
Released 2016-03-14

### Major Changes
- Added set of examples for Juttle tutorial, using github data set [[#85](https://github.com/juttle/juttle-engine/pull/85)]

### Minor Changes
- Bug fixes to examples [[#87](https://github.com/juttle/juttle-engine/pull/87)

## 0.4.0
Released 2016-03-10

### Minor Changes
- The default root path for juttle program files is now the current working directory [[#75](https://github.com/juttle/juttle-engine/pull/75)]
- Updated juttle-viewer to the [0.3.0 release](https://github.com/juttle/juttle-viewer/releases/tag/v0.3.0).
This includes a total reorg of the app layout.
- Make the top level index.html page served by juttle-engine configurable [[#74](https://github.com/juttle/juttle-engine/pull/74)]
- Minor changes to README to aid in understanding the overall ecosystem [[#64](https://github.com/juttle/juttle-engine/pull/64)] [[#72](https://github.com/juttle/juttle-engine/pull/72)]
- Minor changes to example programs [[#65](https://github.com/juttle/juttle-engine/pull/65)] [[#63](https://github.com/juttle/juttle-engine/pull/63)] [[#70](https://github.com/juttle/juttle-engine/pull/70)] [[#71](https://github.com/juttle/juttle-engine/pull/71)] [[#78](https://github.com/juttle/juttle-engine/pull/78)]
- Added automated tests to verify example programs work with latest released versions of dependent modules and adapters [[#58](https://github.com/juttle/juttle-engine/pull/58)] [[#68](https://github.com/juttle/juttle-engine/pull/68)] [[#76](https://github.com/juttle/juttle-engine/pull/76)]

## 0.3.0
Released 2016-02-29

### Major Changes

- Updated juttle to the [0.5.x release](https://github.com/juttle/juttle/releases/tag/v0.5.0). This includes a number of enhancements and breaking language changes.
- Updated juttle-service to the [0.3.0 release](https://github.com/juttle/juttle-service/releases/tag/v0.3.0). This included the following breaking changes:
  - Changed the JSDP wire format to replace all remaining references to 'sink' in JSDP messages to 'view'.
  - Added a /version endpoint that returns version information on the juttle and adapter modules in use.
- Updated all adapters to their latest releases, including compatibility with Juttle 0.5.0 and additional features / changes.
- Added automatic "canary" docker image builds for all changes on all branches. [[#47](https://github.com/juttle/juttle-engine/issues/47)]

### Minor Changes
- Updated juttle-viz to the [0.5.0 release](https://github.com/juttle/juttle-viz/releases/tag/v0.5.0). This included the following major changes:
  - timechart: improve downsampling by changing option name from -display.dataDensity to -downsample and making it a boolean.
- Improved automated testing and test coverage. [[#31](https://github.com/juttle/juttle-engine/issues/31)] [[#34](https://github.com/juttle/juttle-engine/issues/34)] [[#36](https://github.com/juttle/juttle-engine/issues/36)] [[#37](https://github.com/juttle/juttle-engine/issues/37)] [[#46](https://github.com/juttle/juttle-engine/issues/46)]
- Added additional example programs for AWS, cloudwatch, and nginx, and updated other example programs. [[#32](https://github.com/juttle/juttle-engine/issues/32)] [[#33](https://github.com/juttle/juttle-engine/issues/33)] [[#40](https://github.com/juttle/juttle-engine/issues/40)] [[#41](https://github.com/juttle/juttle-engine/issues/41)] [[#56](https://github.com/juttle/juttle-engine/issues/56)] [[#57](https://github.com/juttle/juttle-engine/issues/57)] [[#59](https://github.com/juttle/juttle-engine/issues/59)]

## 0.2.2
Released 2016-02-11

#### Bug Fixes
- Fixed the server so it listens on 0.0.0.0 instead of 127.0.0.1. [[#26](https://github.com/juttle/juttle-engine/issues/26)].
- Updated juttle-service to fix a problem where messages were sometimes dropped when sent over the websocket. [[juttle-service/#38](https://github.com/juttle/juttle-service/pull/38)]
- Updated juttle to work around an npm bug that caused problems on installation. [[juttle/#398](https://github.com/juttle/juttle/pull/398)]

## 0.2.1
Released 2016-02-10

### Bug Fixes
- Added missing entries to package.json so that the bin files get linked into the user PATH as expected. [[#23](https://github.com/juttle/juttle-engine/pull/23)]

## 0.2.0
Released 2016-02-10

### Major Changes
- Initial version of the juttle engine.

## 0.1.0
Released 2016-02-02

### Major Changes
- Initial version of the API service, which was subsequently renamed to [juttle-service](https://github.com/juttle/juttle-service)


gulpfile.js
'use strict';

var eslint = require('gulp-eslint');
var gulp = require('gulp');
var isparta = require('isparta');
var istanbul = require('gulp-istanbul');
var mocha = require('gulp-mocha');

gulp.task('lint', function() {
    return gulp.src([
        'gulpfile.js',
        'bin/*',
        'lib/**/*.js',
        'test/**/*.js'
    ])
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(eslint.failAfterError());
});

gulp.task('instrument', function () {
    return gulp.src([
        'lib/**/*.js',
        'src/**/*.js'
    ])
    .pipe(istanbul({
        includeUntested: true,
        // ES6 Instrumentation
        instrumenter: isparta.Instrumenter
    }))
    .pipe(istanbul.hookRequire());
});

function gulp_test() {
    var timeout;
    var slow;
    var argv = require('minimist')(process.argv.slice(2));

    var tests = [
        'test/**/*.spec.js'
    ];

    // by passing the argument `--sys` you can also run the app tests
    // which require spinning up a browser, by default we do not run
    // the app tests
    if (!argv.sys) {
        tests.push('!test/app/**/*.spec.js');
        tests.push('!test/npm/**/*.spec.js');
        tests.push('!test/examples/**/*.spec.js');
        tests.push('!test/adapters/**/*.spec.js');
        timeout = 20000; // 20s, test timeout
        slow = 15000;    // 15s, slow warning
    } else {
        // downloading and running docker containers can take a bit the first
        // time around
        timeout = 300000; // 5min, test timeout
        slow = 240000;    // 4min, slow warning
    }

    return gulp.src(tests)
    .pipe(mocha({
        log: true,
        timeout: timeout,
        slow: slow,
        reporter: 'spec',
        ui: 'bdd',
        ignoreLeaks: true,
        globals: ['should']
    }));
}

gulp.task('test', function() {
    return gulp_test();
});

gulp.task('test-coverage', ['instrument'], function() {
    var argv = require('minimist')(process.argv.slice(2));
    var thresholds;

    if (argv.sys) {
        thresholds = {
            global: {
                statements: 95,
                branches: 78,
                functions: 97,
                lines: 95
            }
        };
    } else {
        thresholds = {
            global: {
                statements: 96,
                branches: 73,
                functions: 100,
                lines: 96
            }
        };
    }

    return gulp_test()
    .pipe(istanbul.writeReports())
    .pipe(istanbul.enforceThresholds({ thresholds: thresholds }));
});

gulp.task('default', ['lint']);
